### PR TITLE
GEODE-7943 add synchronization to Subscriptions class

### PR DESCRIPTION
**Pull request by jdeppe-pivotal**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by jdeppe-pivotal** [comment]
@ringles @sabbeyPivotal @xtreme-prasath @prettyClouds @jhutchison - for your review

File: geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriptions.java [hunk 1]
```diff
@@ -55,30 +55,37 @@ public boolean exists(Object channelOrPattern, Client client) {
...
-  public void remove(Object channel, Client client) {
-    this.subscriptions.removeIf(subscription -> subscription.isEqualTo(channel, client));
+  public synchronized void remove(Object channel, Client client) {
+    subscriptions.removeIf(subscription -> subscription.isEqualTo(channel, client));
+  }
+
+  /**
+   * @return the total number of all local subscriptions
+   */
+  public int size() {
```
Comment (upthewaterspout) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriptions.java:88]: Technically I think size() needs to be synchronized as well.

Comment (jdeppe-pivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/Subscriptions.java:88]: I've switched to using a `CopyOnWriteArrayList`.

**Comment by jdeppe-pivotal** [comment]
@prettyClouds Good idea! I've switched to using that.

**Comment by jdeppe-pivotal** [comment]
@dschneider-pivotal - Is there any underlying technical difference? I had played around with using a `Vector` which, I think, is practically the same thing. The 'lookup' methods still need to have explicit synchronization although the `add` and `remove` don't. Since there is only a single instance of this class in play, having the synchronization at the method level vs at the level of the backing structure also does not make any difference. Although I could buy the argument that that is just how it is currently being used. I think I just convinced myself... :)

**Comment by dschneider-pivotal** [comment]
why is subscriptionIntegrationTest running out of memory?

**Pull request merged**